### PR TITLE
respect default entity values

### PR DESCRIPTION
**Pull request by taylorfinnell**
See #237

### CODE REVIEW

**Comment by pascalbetz** [comment]
@taylorfinnell why not checking for the presence of the key instead of checking for a _trueish_ value? 
The question i'd ask is "has there been a default assigned" not "is there a default which is true".
Same problem as with assigning the the defaults, actually.

```ruby
if properties[:ruby_default] || properties[:db_type] == 'boolean'
```

=> 

```ruby
if properties.key?(:ruby_default)
```

File: lib/lotus/model/adapters/sql_adapter.rb [hunk 1]
```diff
@@ -281,6 +281,28 @@ def disconnect
 
         private
 
+        # Returns a hash of attributes and their default values
+        #
+        # @api private
+        # @since x.x.x
+        def defaults(collection)
```
Comment (jodosha) [inline lib/lotus/model/adapters/sql_adapter.rb]: This logic can be a private method of `Lotus::Model::Adapters::Sql::Collection`. This is a delegator around a Sequel dataset, so you have direct access to schema informations (via `#schema`).

**Comment by jodosha** [comment]
@taylorfinnell Thanks for starting this.

I think this should honor the mapper first and then the database schema defaults. As of `v0.5.0`, we'll have custom coercers (see #216). That means, a dev can setup their own serialize/deserialize policy for an attribute. I'll give you one silly example to explain the concept.

Let's say we have a `price` column expressed as an integer with cents, so `1€` is stored as `100`. In the Ruby world, the dev wants a float for that attribute, so `1€` is `1.00`. Here's the corresponding code.

```ruby
# migration
column :price, Integer, default: 100
```

```ruby
class Price
  def self.dump(value)
    Integer(value) unless value.nil?
  end

  def self.load(value)
    Float(value) unless value.nil?
  end
end

# mapper
attribute :price, Price
```

Now, with this current implementation, we have a problem:

```ruby
book = Book.new
book = BookRepository.create(book)
book.price # => 100 WRONG
```

We wanted a float, but we have an integer. To solve this issue, we should use `#deserialize_attribute`.

/cc @lotus/core-team @pascalbetz

**Comment by joneslee85** [comment]
@taylorfinnell ping

**Comment by taylorfinnell** [comment]
Oops! Was not getting notifications for this, will work on it this week! Thanks guys

**Comment by DanielBlanco** [comment]
Is there an ETA for this fix? right now I'm doing this:
```
foo = FooRepository.create(foo)
foo = FooRepository.find(foo.id)  # <= Shouldn't need to do this
```
I need it to get default `Sequel.function(:uuid_generate_v4)` value

**Comment by hlegius** [comment]
Guys, any news about it?

**Comment by jodosha** [comment]
@hlegius Nobody's working on it at the moment.

**Comment by joneslee85** [comment]
ping

**Comment by pascalbetz** [comment]
@taylorfinnell posted a question on gitter:

@jodosha Do you have any tips on how to handle the issue mentioned in the first comment here. taylorfinnell/model@0759f90. Working on #237 (lotus/model#239) and have have started to use the deserialize_attribute like you mention but I am not sure how to tell if the user has set their value to nil or if the value is nil and because no value was set and it needs a default

File: lib/lotus/model/adapters/sql_adapter.rb (hunk 1 above)
Comment (taylorfinnell) [inline lib/lotus/model/adapters/sql_adapter.rb]: It looks like #schema is actually on the Sequel connection object not the Sequel dataset object

**Comment by taylorfinnell** [comment]
It looks like attributes that are set to nil don't make it to #create.

Fake migration...
```
  create_table :users do
     column :name, String, default: "Jim"
end
```

```
u = TestUser.new(name: nil)
UserRepo.create(u)
```

In Sql::Collection#insert the `serialized_entity` at that point only contains `id` and no `name`. So I think it would seem reasonable to set the defaults at that point like in the commit @pascalbetz mentioned above

File: lib/lotus/model/adapters/sql/collection.rb [hunk 2]
```diff
@@ -280,6 +283,14 @@ def _serialize(entity)
           def _deserialize(entity)
             @mapped_collection.deserialize([entity]).first
           end
+
+          def _apply_defaults(entity, defaults)
+            defaults.each_pair do |attribute, default|
+              if entity.send(attribute).nil?
```
Comment (taylorfinnell) [inline lib/lotus/model/adapters/sql/collection.rb]: Maybe it's better to serialize the entity and apply the defaults to the hash and then deserialize

Comment (pascalbetz) [inline lib/lotus/model/adapters/sql/collection.rb]: if you serialize first, then you could probably avoid the ```send```?

**Comment by jodosha** [comment]
@taylorfinnell I'm closing this in favor of a new implementation that I'm working on. Thanks!

**Pull request closed without merging**